proposals/0075-precompile-for-secp256r1-sigverify.md

```diff
@@ -0,0 +1,382 @@
+                                               // instruction
+}
+Note: Array<Secp256r1SignatureOffsets> does not contain any length prefixes or
+padding between elements.
+
+struct Secp256r1SignatureOffsets {
+    signature_offset: uint16 LE,              // Offset to signature
+    signature_instruction_index: uint16 LE,   // Instruction index to signature
+    public_key_offset: uint16 LE,             // Offset to public key
+    public_key_instruction_index: uint16 LE,  // Instruction index to  public key
+    message_offset: uint16 LE,                // Offset to start of message data
+    message_length: uint16 LE,                // Size of message data
+    message_instruction_index: uint16 LE,     // Instruction index to message
+}
+```
+
+Up to 8 signatures can be verified. If any of the signatures fail to verify,
+an error must be returned.
+
+In accordance with [SIMD
+0152](https://github.com/solana-foundation/solana-improvement-documents/pull/152)
+the behavior of the program must be as follows:
+
+1. If instruction `data` is empty, return error.
+2. The first byte of `data` is the number of signatures `num_signatures`.
+3. If `num_signatures` is 0, return error.
+4. Expect (enough bytes of `data` for) `num_signatures` instances of
+   `Secp256r1SignatureOffsets`.
+5. For each signature:
+   a. Read `offsets`: an instance of `Secp256r1SignatureOffsets`
+   b. Based on the `offsets`, retrieve `signature`, `public_key`, and
+      `message` bytes. If any of the three fails, return error.
+   c. Invoke the actual `sigverify` function. If it fails, return error.
+
+To retrieve `signature`, `public_key`, and `message`:
+
+1. Get the `instruction_index`-th `instruction_data`
+   - The special value `0xFFFF` means "current instruction"
+   - If the index is invalid, return Error
+2. Return `length` bytes starting from `offset`
+   - If this exceeds the `instruction_data` length, return Error
+
+Note that fields (offsets) can overlap, for example the same public key or
+message can be referred to by multiple instances of `Secp256r1SignatureOffsets`.
+
+If the precompile `verify` function returns any error, the whole transaction
+should fail. Therefore, the type of error is irrelevant and is left as an
+implementation detail.
+
+The instruction processing logic must follow the pseudocode below:
+
+```
+/// `data` is the secp256r1 program's instruction data. `instruction_datas` is
+/// the full slice of instruction datas for all instructions in the transaction,
+/// including the secp256r1 program's instruction data.
+
+/// length_of_data is the length of `data`
+
+/// SERIALIZED_OFFSET_STRUCT_SIZE is the length of the serialized
+/// Secp256r1SignatureOffsets struct
+
+/// SERIALIZED_PUBLIC_KEY_LENGTH and SERIALIZED_SIGNATURE_LENGTH represent the 
+/// length of the serialized public key and signature respectively
+
+function verify() {
+  if length_of_data == 0 {
+    return Error
+  }
+  num_signatures = data[0]
+  if num_signatures > 8 {
```

Comment:
This is unnecessary as it's checked before running precompiles.

Reply:
resolved 👍